wsmux/session.go

```diff
@@ -0,0 +1,351 @@
+package wsmux
+
+import (
+	"bytes"
+	"encoding/binary"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+/*
+TODO: Add ping and pong handlers
+*/
+
+const (
+	defaultQueueSize            = 20
+	defaultStreamQueueSize      = 20
+	defaultKeepAliveInterval    = 30 * time.Second
+	defaultStreamAcceptDeadline = 30 * time.Second
+	deadCheckDuration           = 30 * time.Second
+)
+
+// Session implements net.Listener. Allows creating and acception multiplexed streams over ws
+type Session struct {
+	mu        sync.Mutex
+	streams   map[uint32]*stream
+	streamCh  chan *stream
```

Comment:
how about renaming this `incomingStreams`, or something that indicates that this is streams opened from the other side...